File: node-bindings/classic-node/__test__/path_backups_conformance.ts
import { mkdtemp, writeFile, readFile, readdir, rm } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join, relative, sep } from "node:path";
import * as classic from "../index.js";

/** Observe actual extraction, copy and replacement inside a disposable root. */
export async function observePathBackups(fixture: Record<string, any>): Promise<Record<string, any>> {
  const root = await mkdtemp(join(tmpdir(), "classic-backup-"));
  const local = (path: string) => relative(root, path).split(sep).join("/");
  try {
    const source = join(root, "settings.ini"), log = join(root, "xse.log");
    await writeFile(log, fixture.log); await writeFile(source, Buffer.from(fixture.firstHex, "hex"));
    const manager = new classic.BackupManager(join(root, "backups"));
    const initial = manager.listVersions();
    const version = manager.extractVersionFromXseLog(log), explicit = new classic.XseVersion(fixture.version);
    if (version.fullVersion() !== explicit.fullVersion() || version.sanitized() !== explicit.sanitized() || explicit.toString() !== `XseVersion('${fixture.version}')`) throw new Error("version APIs disagree");
    const created = manager.createBackup(source, version), first = (await readFile(created)).toString("hex");
    await writeFile(source, Buffer.from(fixture.replacementHex, "hex"));
    if (manager.createBackup(source, explicit) !== created) throw new Error("replacement path changed");
    const files: Record<string,string> = {};
    /** Retain every actual file so unexpected backup writes cannot disappear. */
    async function inventory(directory: string): Promise<void> {
      for (const entry of await readdir(directory, { withFileTypes: true })) {
        const path = join(directory, entry.name);
        if (entry.isDirectory()) await inventory(path); else files[local(path)] = (await readFile(path)).toString("hex");
      }
    }
    await inventory(root);
    return { version: version.fullVersion(), sanitized: version.sanitized(), initial, versions: manager.listVersions(), root: local(manager.backupRoot), directory: local(manager.getVersionPath(explicit)), created: local(created), firstHex: first, replacementHex: (await readFile(created)).toString("hex"), files };
  } finally { await rm(root, { recursive: true, force: true }); }
}

File: python-bindings/tests/path_backups_conformance.py
"""Byte-exact versioned backup observations through the Python public API."""

from pathlib import Path
from tempfile import TemporaryDirectory


def observe_path_backups(fixture):
    """Copy arbitrary bytes twice into an owned version directory and read them back."""
    from classic_path import BackupManager, XseVersion

    with TemporaryDirectory(prefix="classic-backup-") as directory:
        root = Path(directory)
        source = root / "settings.ini"
        log = root / "xse.log"
        log.write_bytes(fixture["log"].encode())
        source.write_bytes(bytes.fromhex(fixture["firstHex"]))
        manager = BackupManager(str(root / "backups"))
        initial = manager.list_versions()
        version = manager.extract_version_from_xse_log(str(log))
        explicit = XseVersion(fixture["version"])
        if (version.full_version(), version.sanitized()) != (
            explicit.full_version(),
            explicit.sanitized(),
        ):
            raise ValueError("extracted and explicit versions disagree")
        if (
            repr(explicit) != f"XseVersion('{fixture['version']}')"
            or str(explicit) != fixture["version"]
        ):
            raise ValueError("version representation lost its native value")
        created = Path(manager.create_backup(str(source), version))
        first = created.read_bytes().hex()
        source.write_bytes(bytes.fromhex(fixture["replacementHex"]))
        if Path(manager.create_backup(str(source), explicit)) != created:
            raise ValueError("same-version backup did not replace original path")
        return {
            "version": version.full_version(),
            "sanitized": version.sanitized(),
            "initial": initial,
            "versions": manager.list_versions(),
            "root": Path(manager.backup_root).relative_to(root).as_posix(),
            "directory": Path(manager.get_version_path(explicit))
            .relative_to(root)
            .as_posix(),
            "created": created.relative_to(root).as_posix(),
            "firstHex": first,
            "replacementHex": created.read_bytes().hex(),
            "files": {
                p.relative_to(root).as_posix(): p.read_bytes().hex()
                for p in root.rglob("*")
                if p.is_file()
            },
        }
